File: backend/tests/validation.test.js
/**
 * Validation Test Suite for Student and Alumni Registration Fields
 */

const assert = require('assert');
const app = require('../src/app');
const {
  validateStudentRegistration,
  validateAlumniRegistration,
  validateRegistrationInput,
  isValidEmail,
  isValidMobile,
  isValidPassword
} = require('../src/utils/validation');
const { VALIDATION_MESSAGES } = require('../src/utils/constants');

let passedTests = 0;
let failedTests = 0;

function test(name, fn) {
  try {
    fn();
    console.log(`  ✓ ${name}`);
    passedTests++;
  } catch (err) {
    console.error(`  ✗ ${name}`);
    console.error(`    Error: ${err.message}`);
    failedTests++;
  }
}

async function asyncTest(name, fn) {
  try {
    await fn();
    console.log(`  ✓ ${name}`);
    passedTests++;
  } catch (err) {
    console.error(`  ✗ ${name}`);
    console.error(`    Error: ${err.message}`);
    failedTests++;
  }
}

console.log('====================================================');
console.log('  RUNNING MANUAL INPUT VALIDATION TEST SUITE');
console.log('====================================================\n');

// 1. Email Format Unit Tests
console.log('[1] Email Format Validation Tests');
test('Valid email addresses should pass', () => {
  assert.strictEqual(isValidEmail('student@example.com'), true);
  assert.strictEqual(isValidEmail('john.doe123@university.edu.in'), true);
  assert.strictEqual(isValidEmail('alumni-user@tech.org'), true);
});

test('Invalid email addresses should fail', () => {
  assert.strictEqual(isValidEmail('plainaddress'), false);
  assert.strictEqual(isValidEmail('@no-user.com'), false);
  assert.strictEqual(isValidEmail('user@.com'), false);
  assert.strictEqual(isValidEmail('user@domain'), false);
  assert.strictEqual(isValidEmail(''), false);
  assert.strictEqual(isValidEmail(null), false);
});

// 2. Mobile Number Unit Tests
console.log('\n[2] Mobile Number Validation Tests');
test('Valid mobile numbers should pass', () => {
  assert.strictEqual(isValidMobile('9876543210'), true);
  assert.strictEqual(isValidMobile('+919876543210'), true);
  assert.strictEqual(isValidMobile('8123456789'), true);
  assert.strictEqual(isValidMobile('7000000000'), true);
});

test('Invalid mobile numbers should fail', () => {
  assert.strictEqual(isValidMobile('12345'), false); // Too short
  assert.strictEqual(isValidMobile('1234567890'), false); // Doesn't start with 6-9
  assert.strictEqual(isValidMobile('abcdefghij'), false); // Non-digit
  assert.strictEqual(isValidMobile(''), false);
  assert.strictEqual(isValidMobile(null), false);
});

// 3. Password Criteria Unit Tests
console.log('\n[3] Password Criteria Validation Tests');
test('Valid strong passwords should pass', () => {
  assert.strictEqual(isValidPassword('P@ssword123'), true);
  assert.strictEqual(isValidPassword('Secure#2026'), true);
  assert.strictEqual(isValidPassword('Complex!Pass1'), true);
});

test('Weak passwords missing criteria should fail', () => {
  assert.strictEqual(isValidPassword('short1!'), false); // < 8 chars
  assert.strictEqual(isValidPassword('password123!'), false); // Missing uppercase
  assert.strictEqual(isValidPassword('PASSWORD123!'), false); // Missing lowercase
  assert.strictEqual(isValidPassword('PasswordNoNumber!'), false); // Missing number
  assert.strictEqual(isValidPassword('Password12345'), false); // Missing special char
  assert.strictEqual(isValidPassword(''), false);
});

// 4. Student Registration Validation Tests
console.log('\n[4] Student Registration Field Validation Tests');
test('Valid student registration payload should pass validation', () => {
  const validStudent = {
    fullName: 'Rahul Sharma',
    email: 'rahul.sharma@college.edu',
    mobileNumber: '9876543210',
    password: 'Password@123',
    role: 'student',
    studentId: 'STU2024001',
    department: 'Computer Science',
    graduationYear: 2026
  };
  const result = validateStudentRegistration(validStudent);
  assert.strictEqual(result.isValid, true);
  assert.deepStrictEqual(result.errors, {});
});

test('Student registration missing required fields should return structured error object', () => {
  const emptyStudent = {};
  const result = validateStudentRegistration(emptyStudent);
  assert.strictEqual(result.isValid, false);
  assert.strictEqual(result.errors.fullName, VALIDATION_MESSAGES.FULL_NAME_REQUIRED);
  assert.strictEqual(result.errors.email, VALIDATION_MESSAGES.EMAIL_REQUIRED);
  assert.strictEqual(result.errors.mobileNumber, VALIDATION_MESSAGES.MOBILE_REQUIRED);
  assert.strictEqual(result.errors.password, VALIDATION_MESSAGES.PASSWORD_REQUIRED);
  assert.strictEqual(result.errors.studentId, VALIDATION_MESSAGES.STUDENT_ID_REQUIRED);
  assert.strictEqual(result.errors.department, VALIDATION_MESSAGES.STUDENT_DEPT_REQUIRED);
  assert.strictEqual(result.errors.graduationYear, VALIDATION_MESSAGES.STUDENT_GRAD_YEAR_REQUIRED);
});

test('Student registration with invalid field formats should flag format error messages', () => {
  const invalidStudent = {
    fullName: '1', // Too short/invalid name
    email: 'invalid-email-format',
    mobileNumber: '12345',
    password: 'weak',
    studentId: '$$$',
    department: 'Computer Science',
    graduationYear: 1990 // Past year invalid for student
  };
  const result = validateStudentRegistration(invalidStudent);
  assert.strictEqual(result.isValid, false);
  assert.strictEqual(result.errors.fullName, VALIDATION_MESSAGES.FULL_NAME_INVALID);
  assert.strictEqual(result.errors.email, VALIDATION_MESSAGES.EMAIL_INVALID);
  assert.strictEqual(result.errors.mobileNumber, VALIDATION_MESSAGES.MOBILE_INVALID);
  assert.strictEqual(result.errors.password, VALIDATION_MESSAGES.PASSWORD_INVALID);
  assert.strictEqual(result.errors.studentId, VALIDATION_MESSAGES.STUDENT_ID_INVALID);
  assert.strictEqual(result.errors.graduationYear, VALIDATION_MESSAGES.STUDENT_GRAD_YEAR_INVALID);
});

// 5. Alumni Registration Validation Tests
console.log('\n[5] Alumni Registration Field Validation Tests');
test('Valid alumni registration payload should pass validation', () => {
  const validAlumni = {
    fullName: 'Priya Patel',
    email: 'priya.patel@alumni.org',
    mobileNumber: '9123456789',
    password: 'Alumni@Password2024',
    role: 'alumni',
    graduationYear: 2018,
    department: 'Information Technology',
    company: 'Google',
    designation: 'Senior Software Engineer',
    linkedInProfile: 'https://www.linkedin.com/in/priyapatel'
  };
  const result = validateAlumniRegistration(validAlumni);
  assert.strictEqual(result.isValid, true);
  assert.deepStrictEqual(result.errors, {});
});

test('Alumni registration missing required fields should return structured error object', () => {
  const emptyAlumni = {};
  const result = validateAlumniRegistration(emptyAlumni);
  assert.strictEqual(result.isValid, false);
  assert.strictEqual(result.errors.fullName, VALIDATION_MESSAGES.FULL_NAME_REQUIRED);
  assert.strictEqual(result.errors.email, VALIDATION_MESSAGES.EMAIL_REQUIRED);
  assert.strictEqual(result.errors.mobileNumber, VALIDATION_MESSAGES.MOBILE_REQUIRED);
  assert.strictEqual(result.errors.password, VALIDATION_MESSAGES.PASSWORD_REQUIRED);
  assert.strictEqual(result.errors.graduationYear, VALIDATION_MESSAGES.ALUMNI_GRAD_YEAR_REQUIRED);
  assert.strictEqual(result.errors.department, VALIDATION_MESSAGES.ALUMNI_DEPT_REQUIRED);
  assert.strictEqual(result.errors.company, VALIDATION_MESSAGES.ALUMNI_COMPANY_REQUIRED);
  assert.strictEqual(result.errors.designation, VALIDATION_MESSAGES.ALUMNI_DESIGNATION_REQUIRED);
});

test('Alumni registration with future graduation year or invalid linkedin should fail', () => {
  const invalidAlumni = {
    fullName: 'Anil Kumar',
    email: 'anil@domain.com',
    mobileNumber: '9876543210',
    password: 'Valid123!Pass',
    role: 'alumni',
    graduationYear: 2030, // Future year invalid for alumni
    department: 'Mechanical Engineering',
    company: 'Tesla',
    designation: 'Engineer',
    linkedInProfile: 'not-a-linkedin-url'
  };
  const result = validateAlumniRegistration(invalidAlumni);
  assert.strictEqual(result.isValid, false);
  assert.strictEqual(result.errors.graduationYear, VALIDATION_MESSAGES.ALUMNI_GRAD_YEAR_INVALID);
  assert.strictEqual(result.errors.linkedInProfile, VALIDATION_MESSAGES.LINKEDIN_INVALID);
});

// 6. HTTP API Endpoint Integration Tests
console.log('\n[6] HTTP API Endpoint Server Integration Tests');

async function runHttpTests() {
  const server = app.listen(0);
  const port = server.address().port;
  const baseUrl = `http://localhost:${port}`;

  async function postJson(path, payload) {
    const res = await fetch(`${baseUrl}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload)
    });
    const json = await res.json();
    return { status: res.status, body: json };
  }

  await asyncTest('POST /api/auth/register/student with valid data returns HTTP 201', async () => {
    const response = await postJson('/api/auth/register/student', {
      fullName: 'Aarav Gupta',
      email: 'aarav@college.edu',
      mobileNumber: '9876543210',
      password: 'Student!Pass2026',
      studentId: 'STU1001',
      department: 'ECE',
      graduationYear: 2027
    });
    assert.strictEqual(response.status, 201);
    assert.strictEqual(response.body.success, true);
    assert.strictEqual(response.body.data.email, 'aarav@college.edu');
  });

  await asyncTest('POST /api/auth/register/student with invalid data returns HTTP 400 Bad Request', async () => {
    const response = await postJson('/api/auth/register/student', {
      fullName: 'A',
      email: 'bad-email',
      mobileNumber: '123',
      password: 'pass',
      studentId: '',
      department: '',
      graduationYear: 1980
    });
    assert.strictEqual(response.status, 400);
    assert.strictEqual(response.body.success, false);
    assert.ok(response.body.errors.email);
    assert.ok(response.body.errors.mobileNumber);
    assert.ok(response.body.errors.password);
    assert.ok(response.body.errors.studentId);
  });

  await asyncTest('POST /api/auth/register/alumni with valid data returns HTTP 201', async () => {
    const response = await postJson('/api/auth/register/alumni', {
      fullName: 'Siddharth Varma',
      email: 'sid@alumni.org',
      mobileNumber: '9988776655',
      password: 'Alumni#Secure2024',
      graduationYear: 2020,
      department: 'CSE',
      company: 'Microsoft',
      designation: 'Staff Engineer'
    });
    assert.strictEqual(response.status, 201);
    assert.strictEqual(response.body.success, true);
    assert.strictEqual(response.body.data.company, 'Microsoft');
  });

  await asyncTest('POST /api/auth/register/alumni with invalid data returns HTTP 400 Bad Request', async () => {
    const response = await postJson('/api/auth/register/alumni', {
      fullName: '',
      email: 'invalid-email',
      mobileNumber: '9988776655',
      password: 'Alumni#Secure2024',
      graduationYear: 2040,
      department: '',
      company: '',
      designation: ''
    });
    assert.strictEqual(response.status, 400);
    assert.strictEqual(response.body.success, false);
    assert.ok(response.body.errors.fullName);
    assert.ok(response.body.errors.email);
    assert.ok(response.body.errors.graduationYear);
    assert.ok(response.body.errors.company);
  });

  await asyncTest('POST /api/auth/register (unified) routes correctly based on role', async () => {
    const studentRes = await postJson('/api/auth/register', {
      role: 'student',
      fullName: 'Meera Nair',
      email: 'meera@student.edu',
      mobileNumber: '9112233445',
      password: 'Student#Password1',
      studentId: 'STU2025',
      department: 'Biotech',
      graduationYear: 2026
    });
    assert.strictEqual(studentRes.status, 201);
    assert.strictEqual(studentRes.body.data.role, 'student');
  });

  server.close();

  console.log('\n====================================================');
  console.log(`  RESULTS: ${passedTests} passed, ${failedTests} failed`);
  console.log('====================================================\n');

  if (failedTests > 0) {
    process.exit(1);
  }
}

runHttpTests().catch(err => {
  console.error('Fatal Test Runner Error:', err);
  process.exit(1);
});

File: backend/src/test/java/com/alumni/ValidationTest.java
package com.alumni;

import com.alumni.config.DBConnection;
import com.alumni.server.HttpServerApp;
import com.alumni.service.SocketServer;
import com.alumni.validation.InputValidator;
import com.alumni.validation.ValidationResult;
import org.json.JSONObject;
import org.junit.jupiter.api.*;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.net.HttpURLConnection;
import java.net.Socket;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;

@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class ValidationTest {

    private static HttpServerApp server;
    private static SocketServer socketServer;
    private static int httpPort;
    private static int socketPort;

    @BeforeAll
    public static void setup() throws Exception {
        Assertions.assertTrue(DBConnection.testConnection(), "MySQL connection must be established");

        // Start HTTP test server on dynamic/available port
        httpPort = 5055;
        server = new HttpServerApp(httpPort);
        server.start();

        socketPort = 5056;
        socketServer = new SocketServer(socketPort);
        socketServer.start();

        Thread.sleep(100);
    }

    @AfterAll
    public static void teardown() {
        if (server != null) server.stop();
        if (socketServer != null) socketServer.stop();
    }

    // ==========================================
    // [1] Email Format Validation Tests
    // ==========================================
    @Test
    @Order(1)
    public void testValidEmails() {
        Assertions.assertTrue(InputValidator.isValidEmail("student@university.edu"));
        Assertions.assertTrue(InputValidator.isValidEmail("john.doe@company.com"));
        Assertions.assertTrue(InputValidator.isValidEmail("user+tag@domain.co.in"));
    }

    @Test
    @Order(2)
    public void testInvalidEmails() {
        Assertions.assertFalse(InputValidator.isValidEmail(""));
        Assertions.assertFalse(InputValidator.isValidEmail("not-an-email"));
        Assertions.assertFalse(InputValidator.isValidEmail("missing@domain"));
        Assertions.assertFalse(InputValidator.isValidEmail("@nodomain.com"));
    }

    // ==========================================
    // [2] Mobile Number Validation Tests
    // ==========================================
    @Test
    @Order(3)
    public void testValidMobileNumbers() {
        Assertions.assertTrue(InputValidator.isValidMobile("9876543210"));
        Assertions.assertTrue(InputValidator.isValidMobile("+919876543210"));
        Assertions.assertTrue(InputValidator.isValidMobile("8123456789"));
        Assertions.assertTrue(InputValidator.isValidMobile("7012345678"));
    }

    @Test
    @Order(4)
    public void testInvalidMobileNumbers() {
        Assertions.assertFalse(InputValidator.isValidMobile(""));
        Assertions.assertFalse(InputValidator.isValidMobile("12345"));
        Assertions.assertFalse(InputValidator.isValidMobile("5123456789")); // starts with 5
        Assertions.assertFalse(InputValidator.isValidMobile("abcdefghij"));
    }

    // ==========================================
    // [3] Password Criteria Validation Tests
    // ==========================================
    @Test
    @Order(5)
    public void testValidPasswords() {
        Assertions.assertTrue(InputValidator.isValidPassword("StrongP@ssw0rd"));
        Assertions.assertTrue(InputValidator.isValidPassword("C0mpl3x#P@ss!"));
        Assertions.assertTrue(InputValidator.isValidPassword("Secure123$Auth"));
    }

    @Test
    @Order(6)
    public void testWeakPasswords() {
        Assertions.assertFalse(InputValidator.isValidPassword("short")); // < 8 chars
        Assertions.assertFalse(InputValidator.isValidPassword("alllowercase1!"));
        Assertions.assertFalse(InputValidator.isValidPassword("ALLUPPERCASE1!"));
        Assertions.assertFalse(InputValidator.isValidPassword("NoSpecialChars123"));
        Assertions.assertFalse(InputValidator.isValidPassword("NoNumbers!Pass"));
    }

    // ==========================================
    // [4] Student Registration Field Validation Tests
    // ==========================================
    @Test
    @Order(7)
    public void testValidStudentRegistrationPayload() {
        JSONObject data = new JSONObject();
        data.put("fullName", "Rohan Verma");
        data.put("email", "rohan.verma@college.edu");
        data.put("mobileNumber", "9876543210");
        data.put("password", "Passw0rd!123");
        data.put("studentId", "CS2026001");
        data.put("department", "Computer Engineering");
        data.put("graduationYear", 2026);

        ValidationResult res = InputValidator.validateStudentRegistration(data);
        Assertions.assertTrue(res.isValid());
    }

    @Test
    @Order(8)
    public void testStudentRegistrationMissingFields() {
        JSONObject data = new JSONObject();
        ValidationResult res = InputValidator.validateStudentRegistration(data);

        Assertions.assertFalse(res.isValid());
        Assertions.assertTrue(res.getErrors().containsKey("fullName"));
        Assertions.assertTrue(res.getErrors().containsKey("email"));
        Assertions.assertTrue(res.getErrors().containsKey("mobileNumber"));
        Assertions.assertTrue(res.getErrors().containsKey("password"));
        Assertions.assertTrue(res.getErrors().containsKey("studentId"));
        Assertions.assertTrue(res.getErrors().containsKey("department"));
        Assertions.assertTrue(res.getErrors().containsKey("graduationYear"));
    }

    @Test
    @Order(9)
    public void testStudentRegistrationInvalidFormats() {
        JSONObject data = new JSONObject();
        data.put("fullName", "1"); // too short / numbers
        data.put("email", "invalid-email");
        data.put("mobileNumber", "123");
        data.put("password", "weak");
        data.put("studentId", "ab"); // too short (<3)
        data.put("department", "Computer");
        data.put("graduationYear", 1990); // too far in past for student

        ValidationResult res = InputValidator.validateStudentRegistration(data);
        Assertions.assertFalse(res.isValid());
        Assertions.assertEquals("Please provide a valid email address.", res.getErrors().get("email"));
        Assertions.assertTrue(res.getErrors().containsKey("password"));
        Assertions.assertTrue(res.getErrors().containsKey("studentId"));
    }

    // ==========================================
    // [5] Alumni Registration Field Validation Tests
    // ==========================================
    @Test
    @Order(10)
    public void testValidAlumniRegistrationPayload() {
        JSONObject data = new JSONObject();
        data.put("fullName", "Kavita Rao");
        data.put("email", "kavita.rao@techcorp.com");
        data.put("mobileNumber", "9876543210");
        data.put("password", "Passw0rd!123");
        data.put("department", "Computer Engineering");
        data.put("graduationYear", 2020);
        data.put("company", "TechCorp");
        data.put("designation", "Staff Engineer");
        data.put("linkedInProfile", "https://linkedin.com/in/kavita-rao");

        ValidationResult res = InputValidator.validateAlumniRegistration(data);
        Assertions.assertTrue(res.isValid());
    }

    @Test
    @Order(11)
    public void testAlumniRegistrationMissingFields() {
        JSONObject data = new JSONObject();
        ValidationResult res = InputValidator.validateAlumniRegistration(data);

        Assertions.assertFalse(res.isValid());
        Assertions.assertTrue(res.getErrors().containsKey("fullName"));
        Assertions.assertTrue(res.getErrors().containsKey("email"));
        Assertions.assertTrue(res.getErrors().containsKey("company"));
        Assertions.assertTrue(res.getErrors().containsKey("designation"));
    }

    @Test
    @Order(12)
    public void testAlumniRegistrationInvalidFields() {
        JSONObject data = new JSONObject();
        data.put("fullName", "Kavita Rao");
        data.put("email", "kavita.rao@techcorp.com");
        data.put("mobileNumber", "9876543210");
        data.put("password", "Passw0rd!123");
        data.put("department", "Computer Engineering");
        data.put("graduationYear", 2045); // Future year invalid for alumni
        data.put("company", "TechCorp");
        data.put("designation", "Staff Engineer");
        data.put("linkedInProfile", "not-a-linkedin-url");

        ValidationResult res = InputValidator.validateAlumniRegistration(data);
        Assertions.assertFalse(res.isValid());
        Assertions.assertTrue(res.getErrors().containsKey("graduationYear"));
        Assertions.assertTrue(res.getErrors().containsKey("linkedInProfile"));
    }

    // ==========================================
    // [6] HTTP API Endpoint Server Integration Tests
    // ==========================================
    @Test
    @Order(13)
    public void testPostRegisterStudentSuccess() throws Exception {
        String testEmail = "valid_student_" + System.currentTimeMillis() + "@college.edu";
        JSONObject payload = new JSONObject();
        payload.put("fullName", "Aman Gupta");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543210");
        payload.put("password", "StrongP@ss1");
        payload.put("studentId", "STU" + System.currentTimeMillis());
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2026);

        HttpResult res = sendHttp("POST", "/api/auth/register/student", payload.toString(), "application/json");
        Assertions.assertEquals(201, res.status);
        Assertions.assertTrue(res.json.getBoolean("success"));
        Assertions.assertTrue(res.json.has("id"));

        cleanupUser(res.json.getLong("id"));
    }

    @Test
    @Order(14)
    public void testPostRegisterStudentInvalid() throws Exception {
        JSONObject payload = new JSONObject();
        payload.put("email", "invalid-email");

        HttpResult res = sendHttp("POST", "/api/auth/register/student", payload.toString(), "application/json");
        Assertions.assertEquals(400, res.status);
        Assertions.assertFalse(res.json.getBoolean("success"));
        Assertions.assertTrue(res.json.has("errors"));
    }

    @Test
    @Order(15)
    public void testPostRegisterAlumniSuccess() throws Exception {
        String testEmail = "valid_alumni_" + System.currentTimeMillis() + "@tech.com";
        JSONObject payload = new JSONObject();
        payload.put("fullName", "Ritu Roy");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543211");
        payload.put("password", "StrongP@ss2");
        payload.put("department", "Information Technology");
        payload.put("graduationYear", 2021);
        payload.put("company", "Stripe");
        payload.put("designation", "Software Engineer");

        HttpResult res = sendHttp("POST", "/api/auth/register/alumni", payload.toString(), "application/json");
        Assertions.assertEquals(201, res.status);
        Assertions.assertTrue(res.json.getBoolean("success"));
        Assertions.assertTrue(res.json.has("id"));

        cleanupUser(res.json.getLong("id"));
    }

    @Test
    @Order(16)
    public void testPostRegisterAlumniInvalid() throws Exception {
        JSONObject payload = new JSONObject();
        payload.put("fullName", "Test User");

        HttpResult res = sendHttp("POST", "/api/auth/register/alumni", payload.toString(), "application/json");
        Assertions.assertEquals(400, res.status);
        Assertions.assertFalse(res.json.getBoolean("success"));
        Assertions.assertTrue(res.json.has("errors"));
    }

    @Test
    @Order(17)
    public void testUnifiedRegistrationRouting() throws Exception {
        String testEmail = "unified_student_" + System.currentTimeMillis() + "@college.edu";
        JSONObject payload = new JSONObject();
        payload.put("role", "student");
        payload.put("fullName", "Deepak Shah");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543212");
        payload.put("password", "StrongP@ss3");
        payload.put("studentId", "STU-U-" + (System.currentTimeMillis() % 1000000000L));
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2026);

        HttpResult res = sendHttp("POST", "/api/auth/register", payload.toString(), "application/json");
        Assertions.assertEquals(201, res.status);
        Assertions.assertEquals("student", res.json.getString("role"));

        cleanupUser(res.json.getLong("id"));
    }

    @Test
    @Order(18)
    public void testServletStyleRegistration() throws Exception {
        String testEmail = "servlet_test_" + System.currentTimeMillis() + "@college.edu";
        JSONObject payload = new JSONObject();
        payload.put("role", "student");
        payload.put("fullName", "Servlet User");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543213");
        payload.put("password", "ServletP@ss1");
        payload.put("studentId", "SRV" + System.currentTimeMillis());
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2026);

        HttpResult res = sendHttp("POST", "/api/servlet/register", payload.toString(), "application/json");
        Assertions.assertEquals(201, res.status);
        Assertions.assertTrue(res.json.getBoolean("success"));

        cleanupUser(res.json.getLong("id"));
    }

    @Test
    @Order(19)
    public void testCgiStyleFormRegistration() throws Exception {
        String testEmail = "cgi_test_" + System.currentTimeMillis() + "@college.edu";
        String formBody = "role=student&fullName=CGI+Tester&email=" + testEmail +
                "&mobileNumber=9876543214&password=CgiStrongP%40ss1&studentId=CGI" + System.currentTimeMillis() +
                "&department=Computer+Engineering&graduationYear=2026";

        HttpResult res = sendHttp("POST", "/cgi-bin/register", formBody, "application/x-www-form-urlencoded");
        Assertions.assertEquals(201, res.status);
        Assertions.assertTrue(res.json.getBoolean("success"));

        cleanupUser(res.json.getLong("id"));
    }

    @Test
    @Order(20)
    public void testTcpSocketRegistration() throws Exception {
        String testEmail = "socket_test_" + System.currentTimeMillis() + "@college.edu";
        JSONObject payload = new JSONObject();
        payload.put("role", "student");
        payload.put("fullName", "Socket User");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543215");
        payload.put("password", "SocketP@ss1");
        payload.put("studentId", "SOCK" + System.currentTimeMillis());
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2026);

        try (Socket socket = new Socket("127.0.0.1", socketPort);
             PrintWriter writer = new PrintWriter(socket.getOutputStream(), true, StandardCharsets.UTF_8);
             BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream(), StandardCharsets.UTF_8))) {

            writer.println(payload.toString());
            String responseLine = reader.readLine();
            Assertions.assertNotNull(responseLine);

            JSONObject resp = new JSONObject(responseLine);
            Assertions.assertTrue(resp.getBoolean("success"));
            Assertions.assertTrue(resp.has("id"));

            cleanupUser(resp.getLong("id"));
        }
    }

    @Test
    @Order(21)
    public void testGetRegistrationByEmail() throws Exception {
        String testEmail = "lookup_test_" + System.currentTimeMillis() + "@company.com";
        JSONObject payload = new JSONObject();
        payload.put("fullName", "Maya Sen");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543216");
        payload.put("password", "StrongP@ss4");
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2020);
        payload.put("company", "Salesforce");
        payload.put("designation", "Lead Architect");

        HttpResult regRes = sendHttp("POST", "/api/auth/register/alumni", payload.toString(), "application/json");
        Assertions.assertEquals(201, regRes.status);
        long userId = regRes.json.getLong("id");

        HttpResult getRes = sendHttp("GET", "/api/auth/registrations/" + testEmail, null, null);
        Assertions.assertEquals(200, getRes.status);
        Assertions.assertTrue(getRes.json.getBoolean("success"));
        JSONObject regData = getRes.json.getJSONObject("registration");
        Assertions.assertEquals("Maya Sen", regData.getString("fullName"));
        Assertions.assertEquals("Salesforce", regData.getString("company"));
        Assertions.assertFalse(regData.has("passwordHash"), "Password hash should never be exposed");

        cleanupUser(userId);
    }

    @Test
    @Order(22)
    public void testLoginSuccessAndInvalid() throws Exception {
        // 1. Invalid login
        JSONObject badCreds = new JSONObject();
        badCreds.put("email", "nonexistent.user@example.com");
        badCreds.put("password", "WrongPass@123");
        HttpResult badRes = sendHttp("POST", "/api/auth/login", badCreds.toString(), "application/json");
        Assertions.assertEquals(401, badRes.status);
        Assertions.assertFalse(badRes.json.getBoolean("success"));

        // 2. Register a student and login
        String testEmail = "login_student_" + System.currentTimeMillis() + "@college.edu";
        JSONObject payload = new JSONObject();
        payload.put("fullName", "Login Tester");
        payload.put("email", testEmail);
        payload.put("mobileNumber", "9876543210");
        payload.put("password", "StrongP@ss1");
        payload.put("studentId", "STU" + System.currentTimeMillis());
        payload.put("department", "Computer Engineering");
        payload.put("graduationYear", 2026);

        HttpResult regRes = sendHttp("POST", "/api/auth/register/student", payload.toString(), "application/json");
        Assertions.assertEquals(201, regRes.status);
        long userId = regRes.json.getLong("id");

        // 3. Login with correct credentials
        JSONObject loginPayload = new JSONObject();
        loginPayload.put("email", testEmail);
        loginPayload.put("password", "StrongP@ss1");
        HttpResult loginRes = sendHttp("POST", "/api/auth/login", loginPayload.toString(), "application/json");
        Assertions.assertEquals(200, loginRes.status);
        Assertions.assertTrue(loginRes.json.getBoolean("success"));
        Assertions.assertEquals("student", loginRes.json.getJSONObject("user").getString("role"));
        Assertions.assertEquals("Login Tester", loginRes.json.getJSONObject("user").getString("fullName"));

        cleanupUser(userId);
    }

    @Test
    @Order(23)
    public void testFullNameValidationCases() {
        // Valid names
        Assertions.assertTrue(InputValidator.isValidFullName("Aarav Mehta"));
        Assertions.assertTrue(InputValidator.isValidFullName("Priya Sharma"));
        Assertions.assertTrue(InputValidator.isValidFullName("J. Doe"));
        Assertions.assertTrue(InputValidator.isValidFullName("Mary-Jane Watson"));
        Assertions.assertTrue(InputValidator.isValidFullName("Dr. Bruce Banner"));

        // Invalid names
        Assertions.assertFalse(InputValidator.isValidFullName(null));
        Assertions.assertFalse(InputValidator.isValidFullName(""));
        Assertions.assertFalse(InputValidator.isValidFullName("   "));
        Assertions.assertFalse(InputValidator.isValidFullName("A")); // Too short
        Assertions.assertFalse(InputValidator.isValidFullName("12345")); // Numeric
        Assertions.assertFalse(InputValidator.isValidFullName("Aarav123")); // Contains digits
        Assertions.assertFalse(InputValidator.isValidFullName("@Alex!")); // Special chars
        Assertions.assertFalse(InputValidator.isValidFullName("Aarav   Mehta")); // Double spaces
        Assertions.assertFalse(InputValidator.isValidFullName("A".repeat(55))); // Too long
    }

    @Test
    @Order(24)
    public void testPasswordStrengthEvaluator() {
        // Empty
        InputValidator.PasswordStrength psEmpty = InputValidator.evaluatePasswordStrength("");
        Assertions.assertEquals(0, psEmpty.getScore());
        Assertions.assertEquals("Very Weak", psEmpty.getLevel());
        Assertions.assertFalse(psEmpty.isHasMinLength());

        // Weak (too short, missing criteria)
        InputValidator.PasswordStrength psShort = InputValidator.evaluatePasswordStrength("Ab1!");
        Assertions.assertEquals(1, psShort.getScore());
        Assertions.assertFalse(psShort.isHasMinLength());

        // Missing uppercase
        InputValidator.PasswordStrength psNoUpper = InputValidator.evaluatePasswordStrength("password123!");
        Assertions.assertFalse(psNoUpper.isHasUpperCase());
        Assertions.assertTrue(psNoUpper.isHasMinLength());
        Assertions.assertTrue(psNoUpper.isHasLowerCase());
        Assertions.assertTrue(psNoUpper.isHasDigit());
        Assertions.assertTrue(psNoUpper.isHasSpecialChar());
        Assertions.assertEquals(2, psNoUpper.getScore()); // Fair

        // Missing special char
        InputValidator.PasswordStrength psNoSpecial = InputValidator.evaluatePasswordStrength("Password123");
        Assertions.assertFalse(psNoSpecial.isHasSpecialChar());
        Assertions.assertFalse(InputValidator.isValidPassword("Password123"));

        // Good (all 5 criteria, 8-11 chars)
        InputValidator.PasswordStrength psGood = InputValidator.evaluatePasswordStrength("Passw0rd!");
        Assertions.assertTrue(psGood.isHasMinLength());
        Assertions.assertTrue(psGood.isHasUpperCase());
        Assertions.assertTrue(psGood.isHasLowerCase());
        Assertions.assertTrue(psGood.isHasDigit());
        Assertions.assertTrue(psGood.isHasSpecialChar());
        Assertions.assertEquals(3, psGood.getScore());
        Assertions.assertEquals("Good", psGood.getLevel());
        Assertions.assertTrue(InputValidator.isValidPassword("Passw0rd!"));

        // Strong (all 5 criteria, >= 12 chars)
        InputValidator.PasswordStrength psStrong = InputValidator.evaluatePasswordStrength("Sup3r$ecureP@ssword2026");
        Assertions.assertEquals(4, psStrong.getScore());
        Assertions.assertEquals("Strong", psStrong.getLevel());
        Assertions.assertTrue(InputValidator.isValidPassword("Sup3r$ecureP@ssword2026"));
    }

    @Test
    @Order(25)
    public void testHttpPasswordStrengthEndpoint() throws Exception {
        JSONObject payload = new JSONObject();
        payload.put("password", "Sup3r$tr0ngP@ss1");

        HttpResult res = sendHttp("POST", "/api/auth/password-strength", payload.toString(), "application/json");
        Assertions.assertEquals(200, res.status);
        Assertions.assertTrue(res.json.getBoolean("success"));
        Assertions.assertTrue(res.json.getBoolean("valid"));
        Assertions.assertEquals("Strong", res.json.getJSONObject("strength").getString("level"));
        Assertions.assertEquals(4, res.json.getJSONObject("strength").getInt("score"));
    }

    private static class HttpResult {
        int status;
        JSONObject json;
        String raw;
    }

    private HttpResult sendHttp(String method, String path, String body, String contentType) throws Exception {
        URL url = URI.create("http://127.0.0.1:" + httpPort + path).toURL();
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod(method);
        conn.setConnectTimeout(3000);
        conn.setReadTimeout(3000);

        if (body != null) {
            conn.setDoOutput(true);
            if (contentType != null) conn.setRequestProperty("Content-Type", contentType);
            try (OutputStream os = conn.getOutputStream()) {
                os.write(body.getBytes(StandardCharsets.UTF_8));
            }
        }

        HttpResult res = new HttpResult();
        res.status = conn.getResponseCode();

        try (BufferedReader reader = new BufferedReader(new InputStreamReader(
                res.status >= 400 ? conn.getErrorStream() : conn.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) sb.append(line);
            res.raw = sb.toString();
            try {
                res.json = new JSONObject(res.raw);
            } catch (Exception e) {
                res.json = new JSONObject();
            }
        }
        return res;
    }

    private static void cleanupUser(long userId) {
        try (Connection conn = DBConnection.getConnection();
             PreparedStatement stmt = conn.prepareStatement("DELETE FROM users WHERE id = ?")) {
            stmt.setLong(1, userId);
            stmt.executeUpdate();
        } catch (SQLException ignored) {}
    }
}
